hadoop-ozone/dist/src/main/smoketest/s3/objectputget.robot

```diff
@@ -151,14 +151,34 @@ Incorrect values for end and start offset
                                 Should Be Equal            ${expectedData}            ${actualData}
 
 Zero byte file
-    ${result} =                 Execute AWSS3APICli and checkrc        get-object --bucket ${BUCKET} --key ${PREFIX}/putobject/key=value/zerobyte --range bytes=0-0 /tmp/testfile2.result   255
-                                Should contain             ${result}        InvalidRange
-
-    ${result} =                 Execute AWSS3APICli and checkrc        get-object --bucket ${BUCKET} --key ${PREFIX}/putobject/key=value/zerobyte --range bytes=0-1 /tmp/testfile2.result   255
-                                Should contain             ${result}        InvalidRange
-
-    ${result} =                 Execute AWSS3APICli and checkrc        get-object --bucket ${BUCKET} --key ${PREFIX}/putobject/key=value/zerobyte --range bytes=0-10000 /tmp/testfile2.result   255
-                                Should contain             ${result}        InvalidRange
+    ${result} =			        Execute                    ozone sh bucket info /s3v/${BUCKET}
+    ${linked} = 		        Execute				       echo '${result}' | jq -j '.sourceVolume,"/",.sourceBucket'
+    ${eval} = 			        Evaluate			       "source" in """${linked}"""
+    	      			        IF	${eval} == ${True}
+    ${result} =                 Execute				       ozone sh bucket info ${linked}
+				                END
+    ${fsolayout} =    		    Evaluate    	   	       "OPTIMIZED" in """${result}"""
+
+    ${result} =                 Execute AWSS3APICli and checkrc     get-object --bucket ${BUCKET} --key ${PREFIX}/putobject/key=value/zerobyte --range bytes=0-0 /tmp/testfile2.result   255
+    	      			        IF 	${fsolayout} == ${True}
+                                Should contain              ${result}       NoSuchKey
+				                ELSE
+                                Should contain              ${result}       InvalidRange
+				                END
```

Comment:
@SaketaChalamchala and I just had a brief offline discussion. We think it is strange to return `NoSuchKey` for FSO bucket in this case. The intended behavior should be the same as OBS or LEGACY bucket.

Any idea why this is the case currently? Is it feasible to fix this?

Reply:
This might be happening because all zero byte objects are created as directories in FSO bucket. Would it make sense to also make sure that `put-object --key f1/zero`(without trailing /) creates a file and `put-object --key f1/zero/`(with trailing /) creates a directory in FSO bucket?

```
sh-4.2$ aws s3api --endpoint-url ${ENDPOINT_URL} put-object --bucket fso --key f1/  ## creates f1/ directory.
sh-4.2$ aws s3api --endpoint-url ${ENDPOINT_URL} put-object --bucket fso --key f1/zero --body zero  ## zero is a 0 byte file. creates f1/zero/ dir.
sh-4.2$ aws s3api --endpoint-url ${ENDPOINT_URL} put-object --bucket fso --key f1/test --body test ## test is an 8 byte file. creates f1/test file.

sh-4.2$ ozone sh key list /s3v/fso
[ {
  "volumeName" : "s3v",
  "bucketName" : "fso",
  "name" : "f1/",
  "dataSize" : 0,
  "creationTime" : "2023-04-05T02:11:23.723Z",
  "modificationTime" : "2023-04-05T02:11:23.723Z",
...
}, {
  "volumeName" : "s3v",
  "bucketName" : "fso",
  "name" : "f1/test",
  "dataSize" : 8,
...
}, {
  "volumeName" : "s3v",
  "bucketName" : "fso",
  "name" : "f1/zero/",
  "dataSize" : 0,
...
}
```

Reply:
> @SaketaChalamchala and I just had a brief offline discussion. We think it is strange to return `NoSuchKey` for FSO bucket in this case. The intended behavior should be the same as OBS or LEGACY bucket that returns `InvalidRange`.

I agree FSO should have the same behavior as OBS or LEGACY. However, in order for it to be s3A compatible this is what is needed. Did quite a few tests, and this fixes it. 

> Any idea why this is the case currently? Is it feasible to fix this?

It has to do with internal path normalization for FSO, trailing slashes and what s3A expects as response.

Reply:
> This might be happening because all zero byte objects are created as directories in FSO bucket. 

This is a different issue. It has to do with what s3A expects as a response and what s3G returns. 

> Would it make sense to also make sure that `put-object --key f1/zero`(without trailing /) creates a file and `put-object --key f1/zero/`(with trailing /) creates a directory in FSO bucket?

This could be an idea for additional guard for dir creation. However, I am not sure if it is necessarry because there is already a flag `OZONE_S3G_FSO_DIRECTORY_CREATION_ENABLED` for that. I'd say, right now there is enough room for someone to decide whether a dir or a file should be created (based on [checks](https://github.com/apache/ozone/blob/master/hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java#L216)).

Does this make sense?

Reply:
Yes, it does. Thanks for clarifying.

Reply:
Thanks @mladjan-gadzic